plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/telemetry/otel/OtelBase.kt

```diff
@@ -0,0 +1,243 @@
+// Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.telemetry.otel
+
+import com.intellij.openapi.application.ApplicationInfo
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.platform.diagnostic.telemetry.helpers.useWithoutActiveScope
+import io.opentelemetry.api.common.AttributeKey
+import io.opentelemetry.api.common.Attributes
+import io.opentelemetry.api.trace.Span
+import io.opentelemetry.api.trace.SpanBuilder
+import io.opentelemetry.api.trace.SpanContext
+import io.opentelemetry.api.trace.SpanKind
+import io.opentelemetry.context.Context
+import io.opentelemetry.context.ContextKey
+import io.opentelemetry.context.Scope
+import io.opentelemetry.sdk.trace.ReadWriteSpan
+import kotlinx.coroutines.CoroutineScope
+import software.amazon.awssdk.services.toolkittelemetry.model.AWSProduct
+import software.aws.toolkits.core.utils.error
+import software.aws.toolkits.core.utils.getLogger
+import software.aws.toolkits.core.utils.warn
+import software.aws.toolkits.jetbrains.isDeveloperMode
+import software.aws.toolkits.jetbrains.services.telemetry.PluginResolver
+import java.time.Instant
+import java.util.concurrent.TimeUnit
+import kotlin.coroutines.CoroutineContext
+import kotlin.coroutines.EmptyCoroutineContext
+import com.intellij.platform.diagnostic.telemetry.helpers.use as ijUse
+import com.intellij.platform.diagnostic.telemetry.helpers.useWithScope as ijUseWithScope
+
+val AWS_PRODUCT_CONTEXT_KEY = ContextKey.named<AWSProduct>("pluginDescriptor")
+internal val PLUGIN_ATTRIBUTE_KEY = AttributeKey.stringKey("plugin")
+
+class DefaultSpan(context: Context?, delegate: Span) : BaseSpan<DefaultSpan>(context, delegate)
+
+class DefaultSpanBuilder(delegate: SpanBuilder) : AbstractSpanBuilder<DefaultSpanBuilder, DefaultSpan>(delegate) {
+    override fun doStartSpan() = DefaultSpan(parent, delegate.startSpan())
+}
+
+// temporary; will be generated
+abstract class BaseSpan<SpanType : AbstractBaseSpan<SpanType>>(
+    context: Context?,
+    delegate: Span,
+) : AbstractBaseSpan<SpanType>(context, delegate as ReadWriteSpan)
+
+abstract class AbstractSpanBuilder<
+    BuilderType : AbstractSpanBuilder<BuilderType, SpanType>,
+    SpanType : AbstractBaseSpan<SpanType>,
+    >(
```

Comment:
nit: should the > be on the previous line?

Reply:
can't get formatter to do it